File: services/api/src/utils/response.ts
import { Response } from 'express';

interface ApiResponse<T = unknown> {
  success: boolean;
  message?: string;
  data?: T;
  meta?: Record<string, unknown>;
  error?: { message: string; details?: unknown };
}

export function sendSuccess<T>(res: Response, data: T, message?: string, statusCode = 200, meta?: Record<string, unknown>): void {
  const response: ApiResponse<T> = { success: true };
  if (message) response.message = message;
  response.data = data;
  if (meta) response.meta = meta;
  res.status(statusCode).json(response);
}

export function sendCreated<T>(res: Response, data: T, message = 'Created successfully'): void {
  sendSuccess(res, data, message, 201);
}

export function sendPaginated<T>(res: Response, data: T[], total: number, page: number, limit: number, message?: string): void {
  sendSuccess(res, { items: data, total, page, limit }, message, 200, {
    totalPages: Math.ceil(total / limit),
    hasMore: page * limit < total,
  });
}

export function sendError(res: Response, statusCode: number, message: string, details?: unknown): void {
  const response: ApiResponse = {
    success: false,
    error: { message, ...(details ? { details } : {}) },
  };
  res.status(statusCode).json(response);
}

File: _archive/api-py/app/utils/response.py
from typing import Any, Optional
from fastapi import status
from fastapi.responses import JSONResponse


def success_response(data: Any = None, message: str = 'Success', status_code: int = status.HTTP_200_OK):
    return JSONResponse(
        status_code=status_code,
        content={'success': True, 'message': message, 'data': data},
    )


def created_response(data: Any = None, message: str = 'Created'):
    return success_response(data, message, status.HTTP_201_CREATED)


def paginated_response(items: list, total: int, page: int, limit: int, message: str = 'Success'):
    return JSONResponse(
        status_code=status.HTTP_200_OK,
        content={
            'success': True,
            'message': message,
            'data': {
                'items': items,
                'total': total,
                'page': page,
                'limit': limit,
            },
        },
    )


def error_response(message: str = 'Error', status_code: int = status.HTTP_400_BAD_REQUEST):
    return JSONResponse(
        status_code=status_code,
        content={'success': False, 'message': message},
    )
